test/controllers/supplementary/TransportInformationContainersPageControllerSpec.scala

```diff
@@ -0,0 +1,269 @@
+
+      val cachedData = TransportInformationContainerData(
+        Seq(TransportInformationContainer("M1l3s"), TransportInformationContainer("X4rlz"))
+      )
+      withCaching[TransportInformationContainerData](Some(cachedData), id)
+
+      val result = route(app, getRequest(uri)).get
+      val resultAsString = contentAsString(result)
+
+      status(result) must be(OK)
+
+      resultAsString must include(messages("supplementary.transportInfo.containerId"))
+      resultAsString must include(messages("supplementary.transportInfo.containers.title"))
+      resultAsString must include(messages("supplementary.transportInfo.containerId.title"))
+    }
+  }
+
+  "Additional Information Controller on POST" should {
+
+    "add a container successfully" when {
+
+      "with an empty cache" in {
+        withCaching[TransportInformationContainerData](None, id)
+        val body = Seq(("id", "J0ohn"), addActionURLEncoded)
+
+        val result = route(app, postRequestFormUrlEncoded(uri, body: _*)).get
+
+        status(result) must be(SEE_OTHER)
+      }
+
+      "that does not exist in cache" in {
+        val cachedData = TransportInformationContainerData(Seq(TransportInformationContainer("M1l3s")))
+        withCaching[TransportInformationContainerData](Some(cachedData), id)
+        val body = Seq(("id", "x4rlz"), addActionURLEncoded)
+
+        val result = route(app, postRequestFormUrlEncoded(uri, body: _*)).get
+
+        status(result) must be(SEE_OTHER)
+      }
+    }
+
+    "remove an item successfully" when {
+
+      "exists in cache based on id" in {
+
+        val cachedData = TransportInformationContainerData(
+          Seq(TransportInformationContainer("M1l3s"), TransportInformationContainer("J00hn"))
+        )
+        withCaching[TransportInformationContainerData](Some(cachedData), id)
+
+        val body = removeActionURLEncoded("0")
+        val result = route(app, postRequestFormUrlEncoded(uri, body)).get
+
+        status(result) must be(SEE_OTHER)
+      }
+    }
+
+    "display the form page with an error" when {
+
+      "try to add a container without any data" in {
+
+        withCaching[TransportInformationContainerData](None, id)
+        val body = Seq(("id", ""), addActionURLEncoded)
+
+        val result = route(app, postRequestFormUrlEncoded(uri, body: _*)).get
+        val stringResult = contentAsString(result)
+
+        status(result) must be(BAD_REQUEST)
+        stringResult must include(messages("supplementary.transportInfo.containerId.empty"))
+      }
+
+      "try to save and continue without any data" in {
+
+        val body = Seq(("id", ""), saveAndContinueActionURLEncoded)
+
+        val result = route(app, postRequestFormUrlEncoded(uri, body: _*)).get
+        val stringResult = contentAsString(result)
+
+        status(result) must be(BAD_REQUEST)
+        stringResult must include(messages("supplementary.continue.mandatory"))
+      }
+
+      "try to add a longer id" in {
```

Comment:
please move form based tests into forms and if possible use generators to test all permutation data.

Reply:
First we need to discuss that we want to use generators. We don't have decision about it and you can . find some people that don't want to use it.

Reply:
regardless the tests needs to go into forms specific specs

Reply:
@PatrykRudnicki has a good point on the generators discussion. 
Before I say _yes_ or _no_ to something I normally like to experiment with it.
So I have introduced Generators following loosely what @raghu1978  has made, with the following changes:
1. Unfortunately the `Arbitrary` library expects implicit variables, so I have defined the function on the test method, this way it becomes clearer what we are doing, instead of making our `generator.Generators` a dumping ground of our CDS generative functions.
2. I _have tried_ to make a sensible use of them. After pairing with @MaciejRewera I actually got one previously green test to to red when it shouldn't, i.e, the data generated was incorrect. We are now comfortable that we have fixed this issue :)
3. I actually found the use of Generators interesting but over-complicated for my scenario.